Network/HTTP2/Client/Run.hs

```diff
@@ -236,16 +236,48 @@ sendStreaming
     -> IO (TBQueue StreamingChunk)
 sendStreaming Context{..} strm strmbdy = do
     tbq <- newTBQueueIO 10 -- fixme: hard coding: 10
+    finishedOrCancelled <- newIORef False
     let label = "H2 streaming supporter for stream " ++ show (streamNumber strm)
     forkManagedUnmask threadManager label $ \unmask -> do
         let iface =
                 OutBodyIface
                     { outBodyUnmask = unmask
                     , outBodyPush = \b -> atomically $ writeTBQueue tbq $ StreamingBuilder b NotEndOfStream
-                    , outBodyPushFinal = \b -> atomically $ writeTBQueue tbq $ StreamingBuilder b (EndOfStream Nothing)
+                    , outBodyPushFinal = \b -> do
+                        atomicWriteIORef finishedOrCancelled True
+                        atomically $ writeTBQueue tbq $ StreamingBuilder b (EndOfStream Nothing)
+                        atomically $ writeTBQueue tbq $ StreamingFinished Nothing
```

Comment:
I'm a bit worried here about atomicity; what happens if an exception is thrown to the thread that calls `outBodyPushFinal`, and we do update `finishedOrCancelled` but don't enqueue the rest? (Note that such a call would run in the context of the calling code, not in the context of a `http2` created thread.) Or indeed, enqueue the `StreamingBuilder` but not the `StreamingFinished`? Perhaps it would be better to make `finishedOrCancelled` an `TVar` rather than an `IORef`, and make this a single transaction with three statements in it.